### ISSUE TITLE
can't be used with Typescript anymore

### ISSUE THREAD

**Issue by pocesar** [keichi/binary-parser#112]
all dynamically generated functions are gone from the typings (`[u]int[\d]`)

![image](https://user-images.githubusercontent.com/461084/58929908-9c7a8a80-872f-11e9-9da8-6d1faf250e9e.png)

when using Typescript, you have to create all your methods, in this case, create a function that serves as a proxy for the:

```ts
    Object.keys(PRIMITIVE_TYPES).forEach(type => {
      Parser.prototype[type.toLowerCase()] = function(varName, options) {
        return this.setNextParser(type.toLowerCase(), varName, options);
      };

      var typeWithoutEndian = type.replace(/BE|LE/, '').toLowerCase();
      if (!(typeWithoutEndian in Parser.prototype)) {
        Parser.prototype[typeWithoutEndian] = function(varName, options) {
          return this[typeWithoutEndian + this.endian](varName, options);
        };
      }
    });
```

then delegate inside each method:

```ts
  public uint8(varName: string, options: ParseOptions) {
    return this.setNextParser('uint8', varName, options);
  }
```

it's verbose, since you're coming from the JS version and trying to port, there's no way around this. you either type everything, or you may need to rely on JSDOC to annotate missing / virtual methods. 

you're also missing releasing the JS version on npm as well, people that aren't using Typescript won't be able to use version 1.3.3

![image](https://user-images.githubusercontent.com/461084/58930166-bd8fab00-8730-11e9-956f-61dac87659b4.png)

there's also a hardcoded devDependency as a dependency

**Comment by Toxicable** [comment]
most of these issues have been fixed in master, it just needs a release

**Comment by keichi** [comment]
Hi,

> when using Typescript, you have to create all your methods, in this case, create a function that serves as a proxy for the:

This has been addressed in #101 and will be shipped in the next release. We have statically declared all parser methods.

> there's also a hardcoded devDependency as a dependency

This has been fixed in #108.

> you're also missing releasing the JS version on npm as well, people that aren't using Typescript won't be able to use version 1.3.3

I just installed 1.3.3 from npm, but I do see the compiled JS files under `dist/`. Can you describe how you installed it?

```
$ npm install binary-parser@1.3.3
npm WARN saveError ENOENT: no such file or directory, open '/Users/keichi/Projects/scratch/package.json'
npm WARN enoent ENOENT: no such file or directory, open '/Users/keichi/Projects/scratch/package.json'
npm WARN mocha-junit-reporter@1.22.0 requires a peer of mocha@>=2.2.5 but none is installed. You must install peer dependencies yourself.
npm WARN scratch No description
npm WARN scratch No repository field.
npm WARN scratch No README data
npm WARN scratch No license field.

+ binary-parser@1.3.3
updated 1 package and audited 42 packages in 0.672s
found 0 vulnerabilities

$ tree node_modules/binary-parser/
node_modules/binary-parser/
├── LICENSE.md
├── README.md
├── dist
│   ├── binary_parser.d.ts
│   ├── binary_parser.d.ts.map
│   ├── binary_parser.js
│   ├── binary_parser.js.map
│   ├── context.d.ts
│   ├── context.d.ts.map
│   ├── context.js
│   └── context.js.map
├── lib
│   ├── binary_parser.ts
│   └── context.ts
├── package.json
└── tsconfig.json

2 directories, 14 files
```

**Comment by pocesar** [comment]
about the npm install, it was an error on my part (maybe bad cache?), it has a `dist` folder now. 

another suggestion: I started modifying the `@types/binary-parser` to make it even more strict, since it already provides a generic interface for the object shape, it's currently better than the current Typescript sources. ie: you can use `Parser<{ member: number }>` and cast using `as`. while you're at it, provide the merging of declaration when creating the object, something like:

```ts
class Parser<T> {
   uint8(): Parser<T & somethingToMerge> {
   }
}
```
you can see the current typings in https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/binary-parser/index.d.ts

**Comment by keichi** [comment]
Thanks for the suggestion, we should base our interface design on`@types/binary-parser`.

**Comment by keichi** [comment]
Just released v1.4.0, which includes the fixes I mentioned above. I temporarily excluded the TypeScript files from the package since I found out some of the typings were incomplete (#116).
